### PR TITLE
PR: Add documentation for Line Profiler plugin

### LINKED ISSUE #273: New doc request: Line Profiler plugin
# Issue Report
It would be nice to have a docs page explaining how to use the [Spyder Line Profiler plugin](https://github.com/spyder-ide/spyder-line-profiler/).

## Issue Description
- [ ] Create a page in the plugins folder explaining how to use the Line Profiler plugin
- [ ] Add images that showcase the Line profiler plugin
- [ ] Verify that the docs build correctly

### Where is the relevant portion of the documentation located?
In the plugins section.

### Other information

* Is the problem specific to documentation for a particular version of Spyder?
To `4.x`

* Are you planning on submitting a Github pull request with a suitable change?
Yes, I will attempt a PR shortly.

* Anything else you'd like to add?
This issue is part of the SciPy 2021 sprint.

### PR DESCRIPTION
**Pull request by vinisalazar**
Hi,

this PR is part of the SciPy 2021 sprint.

## Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)
* [x] Based your PR on the latest version of the correct branch (master or 4.x)
* [x] Checked your writing carefully for correct English spelling, grammar, etc
* [x] Described your changes and the motivation for them below
* [x] Noted what issue(s) this pull request resolves, creating one if needed

## Description of Changes
  - Add page for lineprofile plugin
  - Add directory for the lineprofiler plugin images
  - Add images showcasing the use of the plugin

## Issue(s) Resolved

Fix #273

### CODE REVIEW

File: doc/plugins/lineprofiler.rst
Comment (juanis2112) [inline doc/plugins/lineprofiler.rst:10]:
```suggestion
The line_profiler will profile the time that individual lines of code take to execute.
```
This was probably from the other PR and I don't think it applies. Maybe here it would be worth describing what a line profiler does. I added that part from the link you included but feel free to change it.

File: doc/plugins/lineprofiler.rst
Comment (juanis2112) [inline doc/plugins/lineprofiler.rst:11]:
```suggestion
```

File: doc/plugins/lineprofiler.rst
Comment (juanis2112) [inline doc/plugins/lineprofiler.rst:49]: Maybe add something here about the path of the profiler saying that the path displayed there is the path of the file that we are profiling?  @ccordoba12 what do you think?

**Review by juanis2112 — COMMENTED** [review]
This looks great. Some minor suggestions. BTW great gifs!!
I will add @jitseniesen as reviewer since he is the maintainer of the plugin just to know if we are missing something.

**Comment by juanis2112** [comment]
@jitseniesen we have now docs for our Line Profiler thanks to @vinisalazar who worked on it during our Scipy sprints! Please let us know what you think about this PR and if there is something we are missing.

File: doc/plugins/lineprofiler.rst
Comment (vinisalazar) [reply doc/plugins/lineprofiler.rst]: Thank you for pointing this out @juanis2112! Yes, I did copy the `.rst` from my previous PR for this, but overlooked this.

File: doc/plugins/lineprofiler.rst
Comment (vinisalazar) [reply doc/plugins/lineprofiler.rst:48]: I'll wait for feedback before updating this section.

**Review by vinisalazar — COMMENTED** [review]
Thank you for the review @juanis2112 ! I applied the request changes and will stand by to wait for any other suggestions.

File: doc/plugins/lineprofiler.rst
Comment (ccordoba12) [inline doc/plugins/lineprofiler.rst:5]:
```suggestion
**Spyder-Line-Profiler** is a plugin to run the Python `line profiler <https://pypi.python.org/pypi/line_profiler>`_.
This package profiles the time that individual lines of code take to execute.
```

- There's no need to join `line` and `profiler` by an underscore.
- I also moved the line after the screenshot here because I think it makes more sense.

File: doc/plugins/lineprofiler.rst
Comment (ccordoba12) [inline doc/plugins/lineprofiler.rst:10]:
```suggestion
```

I'd suggest to remove this line from here to move it to the first sentence.

File: doc/plugins/lineprofiler.rst
Comment (ccordoba12) [inline doc/plugins/lineprofiler.rst:16]:
```suggestion
If you installed Spyder using conda, the best way to install Spyder-line-profiler is to run the following command in your terminal or Anaconda prompt on Windows:
```

Correct typo.

File: doc/plugins/lineprofiler.rst
Comment (ccordoba12) [inline doc/plugins/lineprofiler.rst:38]:
```suggestion
Afterwards, your file will be profiled line by line.
```

- Give a better start to this sentence.
- Avoid repeating `your script`, which was mentioned in the previous sentence.
- Mention again what the plugin does (just for clarity).

File: doc/plugins/lineprofiler.rst
Comment (ccordoba12) [reply doc/plugins/lineprofiler.rst:48]: Good suggestion @juanis2112! I agree with it.

**Review by ccordoba12 — COMMENTED** [review]
Thanks a lot @vinisalazar for your help with this!

This looks pretty good for me, I just left some minor suggestions for you.

File: doc/plugins/lineprofiler.rst
Comment (vinisalazar) [reply doc/plugins/lineprofiler.rst:48]: Fixed by b698589.

**Comment by vinisalazar** [comment]
Once again, thank you for the review. I've applied the requested changes.

**Review by jitseniesen — APPROVED** [review]
Thanks for writing these docs. They generally look fine to me, so I'm happy to have it merged as it. I have two suggestions that you may want to address now or in a future PR.

Firstly, I wonder whether a slight reordering makes sense. After "You will see it then as a tab next to the 'Files' tab", first say that you have to add the `@profile` decorator (with the GIF) and then say that you can either select and run the script from the pane or from the Run menu. This is the order in which users will typically do things (I expect).

Secondly, perhaps add something on how to interpret the results from the profile.

**Comment by vinisalazar** [comment]
Hi @jitseniesen,

thank you for the review. I've applied the requested changes, and added you as a co-author to the commits. I hope that's okay. If not, let me know and I can amend the commit messages.

Best,
Vini

File: doc/plugins/lineprofiler.rst
Comment (jitseniesen) [inline doc/plugins/lineprofiler.rst:48]: I would remove this line ("It's also possible to ...") because it repeats what you say two sentences above, or you can reformulate it, for instance: "After running the Line Profiler, either from the button in the Pane or from the :menuselection:`Run` menu, the results are shown in the Line Profiler pane."

File: doc/plugins/lineprofiler.rst
Comment (jitseniesen) [inline doc/plugins/lineprofiler.rst:57]:
```suggestion
   * ``Time (ms):`` time spent running the line in total for all hits, in milliseconds.
```
Milli- has two Ls in English.

File: doc/plugins/lineprofiler.rst
Comment (jitseniesen) [inline doc/plugins/lineprofiler.rst:58]:
```suggestion
   * ``Per hit (ms):`` average time spent per hit, in milliseconds.
```
As above.

**Review by jitseniesen — CHANGES_REQUESTED** [review]
It looks good now apart from a misspelling. It wasn't necessary to add me as an author but I don't mind so it's probably easiest now to leave it as it is.

File: doc/plugins/lineprofiler.rst
Comment (vinisalazar) [reply doc/plugins/lineprofiler.rst]: Resolved by d9a8d97.

**Comment by vinisalazar** [comment]
@jitseniesen once again, thank you. I hope the PR is good to go now.

Best,
Vini

**Review by jitseniesen — APPROVED** [review]
Looks good, thanks a lot for your work!

**Comment by jitseniesen** [comment]
@juanis2112 I approved the PR. I think the next step is to press the button "Approve and run" under "1 workflow awaiting approval" but I'm hesitant to do so because I don't know what it means. Please explain and/or press the button yourself.

**Review by juanis2112 — APPROVED** [review]
Sorry it took me so long to get back at this!  I was busy on the workshops we just released! This looks great to me! Thanks @vinisalazar for working on this  and @jitseniesen for the review.

**Comment by juanis2112** [comment]
@ccordoba12 I think you can merge this.

File: doc/plugins/lineprofiler.rst
Comment (CAM-Gerlach) [inline doc/plugins/lineprofiler.rst:62]: I suggest you be more specific here. I'm not sure what precise a "stronger" color is supposed to mean, and thus many users likely won't either. Do you mean more saturated? Brighter? Darker? Or something more subjective?

**Review by CAM-Gerlach — CHANGES_REQUESTED** [review]
Excellent job @vinisalazar ! Very well done overall and a great addition to our docs. I just have a few writing tweaks and minor nitpicks, all as suggestions that you can apply in one batch, in addition to one comment.

Also, I manually fixed the issues it would have caught this time, but don't forget to install Pre-Commit as required by the Contributing Guide you indicated you read. I'm not sure why the checks aren't running (maybe the first time contributor run request expired?), but hopefully pushing these fixes will start things up.

Your GIFs were well done, but both had a very visible moire pattern that made them very hard to see (example below), at least on non-hi-DPI screens (as both your and @juanis2112 's screens happen hiDPI and display the images at full size, the effect is much more subtle). This is caused by something in your pipeline adding color dithering (the pattern of small dots you can see when viewing the GIF full size), coupled with the overly large resolution of the GIFs resulting in them being downscaled, triggering the moire pattern. If you still have the original source video, the best solution is to re-create the GIFs without dithering enabled. If not, and it is too much trouble to redo them, downsampling them yourself should greatly reduce the effect (you may have to try different methods, e.g. lanczos, bicubic or nearest neighbor). This is also needed so they have a more reasonable file size and match the other GIFs in this repo, all of which have a resolution one quarter or less of these (i.e. 720 x 450).

image

Your images looked good, but they are much larger than they need to be. Please downscale them to 1440 x 900 , one quarter of their current resolution and the standard size of images in this repo, to avoid consuming undue repository space and download bandwidth, and run optipng on them, which resulted in a further 30-50% savings for free (you can install it with e.g. `brew install optipng`, then from the root of the repo, run `optipng doc/images/lineprofiler/*.png`).

In order to ensure the old, larger versions of the images and GIFs don't permanently increase the size of this repo, you'll need to fixup the updated assets into the original commits via `git rebase -i`. If you're not comfortable with advanced git fu, or any of these other steps, we can do all of them for you once this PR is otherwise ready to merge, other than regenerating/redoing the GIFs (though if necessary, we can resize the existing ones, though this will result in a significant loss of quality).

Thanks again for your hard work on this, and looking forward to seeing this in Spyder!

File: doc/plugins/lineprofiler.rst
Comment (CAM-Gerlach) [reply doc/plugins/lineprofiler.rst]: Looks like this was resolved, thanks.

**Comment by ccordoba12** [comment]
@CAM-Gerlach, could we merge this one as it is? Since it's already approved by @jitseniesen and @juanis2112, I think it's good to go.

Then you could open a new PR to make the changes you suggested and @juanis2112 could redo the gifs in another one.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 I'm sorry for my extended absence causing these issues. Unfortunately, unlike PR #282 where I suggested exactly that, that's not really possible to do here, since other than one small comment and a few trivial technical/style changes that can simply be batch-applied in one commit and don't make any non-trivial changes to the content (and it appears another commit is needed anyway to trigger the CIs to run) my suggestions are all related to the images and GIFs being four times the nominal resolution and unoptimized, and the latter furthermore having crippling artifacts on non-hiDPI screens (they look bad enough in the screenshot but much worse in motion). If this PR is merged, that's a substantial chunk of repo space and clone/sync bandwidth we will never be able to get back, because it will be impossible to delete the blobs from everyone's repos without re-writing history on `master`.

It would be easier and substantially better quality for @vinisalazar to re-generate the GIFs from the source video without the dithering, but I believe it should be possible for me to fix the worst of the problem using the GIFs as submitted. Thus, if @vinisalazar prefers, given the PR is otherwise ready to go, it would be straightforward for me to fix/optimize and rebase the existing images/GIFs on their branch for them, like I've done for every previous docs PR of this nature, and then it would be ready to merge with nothing further needed from anyone else.

File: doc/plugins/lineprofiler.rst
Comment (vinisalazar) [reply doc/plugins/lineprofiler.rst:62]: Thank you for your comment. This was quoted directly from the plugin's [README page](https://github.com/spyder-ide/spyder-line-profiler/blob/master/README.md#usage). It'd be better to ask the developers.

Comment (CAM-Gerlach) [reply doc/plugins/lineprofiler.rst:70]: Okay, thanks! That's fine then.

**Comment by CAM-Gerlach** [comment]
Thanks so much for your great work on this, @vinisalazar . Other than the scaling/optimization issues with the images/GIFs which I can fix for you, everything looks great. I'll go ahead and do that now, and then this should be ready to merge. Cheers!

**Comment by CAM-Gerlach** [comment]
Oh and by the way, great job on your detailed and well structured commit messages. I'll need to do a bit of light squashing to consolidate some of the smaller review commits and reduce the chance of merge conflicts when rebasing in the GIF/image changes, but I'll make sure to preserve your commit structure and messages when I do so. Thanks!

**Review by CAM-Gerlach — APPROVED** [review]
It took some work, but I was able to fix the image and GIF issues and reduce their size substantially (albeit not as efficiently as if done from the original source) and rebased the changes in. Other than that, looks great @vinisalazar , and so with approval from all involved I'll go ahead and merge. Thanks!

**Comment by vinisalazar** [comment]
@CAM-Gerlach thanks a lot for your work in this PR :)

**Comment by CAM-Gerlach** [comment]
You're most welcome, but I really should be thanking you, as you did all the real work, along with @juanis2112 , @jitseniesen and @ccordoba12 .

**Pull request merged**
